### PR TITLE
Disallow declaring reserved @@variable names

### PR DESCRIPTION
**Pull request by robverschoor**
### Description

Now that variables named `@@myvar` are supported (see [PR #3073](https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/3073)), we should also address the issue of being able to declare a user-defined variable with the same name as a predefined global `@@` variable, for example `@@SPID`.
Babelfish has always allowed such declarations even though referencing the declared variable would actually reference the predefined global `@@variable`, thus producing a wrong result. For example:

```
1> declare @@spid int = 123
2> select @@spid
3> go
spid
-----------
      20902     <===== WRONG!
```
      
This fix does not allow predefined `@@` global variables to be used as user-defined variables or parameters, and will raise a syntax error as expected in T-SQL.
A special case is the `@@PGERROR` variable which was introduced by Babelfish: when declaring a variable with this name, the error message will additionally point out that `@@PGERROR` is a reserved name in Babelfish. 

Signed-off-by: Rob Verschoor [rcv@amazon.com](mailto:rcv@amazon.com)

### Issues Resolved

BABEL-5388: Disallow declaring reserved @@variable names

### Test Scenarios Covered ###
- Use case based - Yes

- Boundary conditions - Yes

- Arbitrary inputs - N/A

- Negative test cases - Yes

- Minor version upgrade tests - Yes

- Major version upgrade tests - N/A

- Performance tests - N/A

- Tooling impact - N/A

- Client tests - N/A

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: test/JDBC/expected/declare_atatglobalvars_upgrade-before-16_5-vu-verify.out [hunk 1]
```diff
@@ -0,0 +1,16 @@
+exec p1_declare_atatglobalvars_upgrade
```
Comment (KushaalShroff) [inline test/JDBC/expected/declare_atatglobalvars_upgrade-before-16_5-vu-verify.out]: Why does this not exist post upgrade? Ideally the prepare script would run on 16.4 and when we upgrade to 16.5/16.6 then the object should still exist

Comment (robverschoor) [reply test/JDBC/expected/declare_atatglobalvars_upgrade-before-16_5-vu-verify.out]: The upgrade would fail as the object cannot be recreated anymore. With the fix in this PR, the failure happens when the object is created.

Comment (KushaalShroff) [reply test/JDBC/expected/declare_atatglobalvars_upgrade-before-16_5-vu-verify.out]: Object wont be recreated from 16.4 to 16.5/16.6, even if they did, the restore happens from psql endpoint and ANTLR does not come into picture. Dump/restore happens only for Major version upgrade.

File: test/JDBC/expected/declare_atatglobalvars_upgrade-before-16_5-vu-verify.out [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+go
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: Incorrect syntax near '@@max_precision'.)~~
+
+exec p1_declare_atatglobalvars_upgrade
+go
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: Incorrect syntax near '@@max_precision'.)~~
```
Comment (KushaalShroff) [inline test/JDBC/expected/declare_atatglobalvars_upgrade-before-16_5-vu-verify.out:11]: So we have a case where previous functions will stop working are we expecting the customer experience to be the same?
Can we take the risk of assuming that there wouldnt be such dependent objects in the first place since these are not valid TSQL?

File: test/JDBC/input/declare_atatglobalvars-vu-verify.sql [hunk 3]
```diff
@@ -0,0 +1,212 @@
+declare @@cursor_rows int
```
Comment (KushaalShroff) [inline test/JDBC/input/declare_atatglobalvars-vu-verify.sql:1]: declare_atatglobalvars For this file , I think we need to add it to the latest/schedule file as well

**Review by KushaalShroff — CHANGES_REQUESTED** [review]

File: test/JDBC/expected/declare_atatglobalvars_upgrade-before-16_5-vu-verify.out (hunk 2 above)
Comment (robverschoor) [reply test/JDBC/expected/declare_atatglobalvars_upgrade-before-16_5-vu-verify.out:11]: If such objects had existed, which would be highly unlikely since they could not have been created in SQL Server in the first place, then they would have produced incorrect results (see PR description). That was wrong, and this fix stops it from happening. This is an extreme edge scenario and would only occur if customers had created an object using syntax that was invalid in SQL Server.

File: test/JDBC/input/declare_atatglobalvars-vu-verify.sql (hunk 3 above)
Comment (robverschoor) [reply test/JDBC/input/declare_atatglobalvars-vu-verify.sql:1]: done

**Review by KushaalShroff — APPROVED** [review]
